simulation/traffic_simulator/src/traffic_lights/traffic_light_state.cpp

```diff
@@ -0,0 +1,160 @@
+
+namespace traffic_simulator
+{
+std::istream & operator>>(std::istream & is, TrafficLightColor & datum)
+{
+  std::string value;
+
+  is >> value;
+
+  static const std::unordered_map<std::string, TrafficLightColor> conversions{
+    // clang-format off
+    std::make_pair("none",   TrafficLightColor::NONE),
+    std::make_pair("red",    TrafficLightColor::RED),
+    std::make_pair("green",  TrafficLightColor::GREEN),
+    std::make_pair("yellow", TrafficLightColor::YELLOW),
+    // clang-format on
+  };
+
+  try {
+    datum = conversions.at(value);
+  } catch (const std::out_of_range &) {
+    std::stringstream what;
+    what << "An invalid value '" << value
+         << "' was specified for type 'traffic_simulator::TrafficLightColor'.";
+    throw std::runtime_error(what.str());
+  }
+
+  return is;
+}
+
+std::ostream & operator<<(std::ostream & os, const TrafficLightColor & datum)
+{
+  // clang-format off
+  switch (datum) {
+    case TrafficLightColor::NONE:   return os << "none";
+    case TrafficLightColor::RED:    return os << "red";
+    case TrafficLightColor::GREEN:  return os << "green";
+    case TrafficLightColor::YELLOW: return os << "yellow";
+  }
+  // clang-format on
+}
+
+autoware_perception_msgs::msg::LampState makeLampState(const TrafficLightColor & datum)
+{
+  autoware_perception_msgs::msg::LampState lamp_state;
+  {
+    lamp_state.confidence = 1.0;
+
+    switch (datum) {
+      case TrafficLightColor::RED:
+        lamp_state.type = autoware_perception_msgs::msg::LampState::RED;
+        break;
+
+      case TrafficLightColor::GREEN:
+        lamp_state.type = autoware_perception_msgs::msg::LampState::GREEN;
+        break;
+
+      case TrafficLightColor::YELLOW:
+        lamp_state.type = autoware_perception_msgs::msg::LampState::YELLOW;
+        break;
+
+      default:
+        std::stringstream what;
+        what << "Casting TrafficLightColor::" << datum
+             << " to autoware_perception_msgs::msg::LampState is not supported.";
+        throw std::out_of_range(what.str());
+    }
+  }
+
+  return lamp_state;
+}
+
+std::istream & operator>>(std::istream & is, TrafficLightArrow & datum)
+{
+  std::string value;
+
+  is >> value;
+
+  static const std::unordered_map<std::string, TrafficLightArrow> conversions{
+    // clang-format off
+    std::make_pair("none",     TrafficLightArrow::NONE),
+    std::make_pair("straight", TrafficLightArrow::STRAIGHT),
+    std::make_pair("left",     TrafficLightArrow::LEFT),
+    std::make_pair("right",    TrafficLightArrow::RIGHT),
+    // clang-format on
+  };
+
+  try {
+    datum = conversions.at(value);
+  } catch (const std::out_of_range &) {
+    std::stringstream what;
+    what << "An invalid value '" << value
+         << "' was specified for type 'traffic_simulator::TrafficLightArrow'.";
+    throw std::runtime_error(what.str());
+  }
+
+  return is;
+}
+
+std::ostream & operator<<(std::ostream & os, const TrafficLightArrow & datum)
+{
+  // clang-format off
```

Comment:
Please avoid to use clang-format off and on if possible.

Reply:
I used it exceptionally when I wanted the code to look tabular.